_pytest/python.py

```diff
@@ -293,11 +293,11 @@ def pytest_pycollect_makeitem(collector, name, obj):
     elif collector.istestfunction(obj, name):
         # mock seems to store unbound methods (issue473), normalize it
         obj = getattr(obj, "__func__", obj)
-        if not isfunction(obj):
+        if not isfunction(get_real_func(obj)):
             collector.warn(code="C2", message=
                 "cannot collect %r because it is not a function."
                 % name, )
-        if getattr(obj, "__test__", True):
+        elif getattr(obj, "__test__", True):
```

Comment:
No my tests do not pass. py.test collects and runs the test and fails with an internal error as before. 

This line is not checking for explicit __test__ attributes. The if statement only has an effect if __test__ is set to False. Because it's defaulting the getattr to True.